graphviz/quoting.py

```diff
@@ -37,8 +37,93 @@
+    See: https://graphviz.org/doc/info/lang.html
+
+        node_id     :   ID [ port ]
+
+        port        :   ':' ID [ ':' compass_pt ]
+                    |   ':' compass_pt
+
+        compass_pt  :   (n | ne | e | se | s | sw | w | nw | c | _)
+
+    Note that this syntax element is used in both node and edge statements.
+    That is, you can declare a node in the following ways:
+
+        digraph {
+            node1
+            node1:port1
+            node1:port1:nw
+        }
+
+    ...as far as I can tell, the port and compass are ignored, so those 3
+    node statements are equivalent.
+    You can confirm this yourself (somewhat) at the commandline:
+
+        $ echo "digraph { a:b }" | dot
+        digraph {
+            graph [bb="0,0,54,36"];
+            node [label="\N"];
+            a   [height=0.5,
+                pos="27,18",
+                width=0.75];
+        }
+
+        $ echo "digraph { a:b; a:b -> c }" | dot
+        Warning: node a, port b unrecognized
+        digraph {
+            graph [bb="0,0,54,108"];
+            node [label="\N"];
+            a   [height=0.5,
+                pos="27,90",
+                width=0.75];
+            c   [height=0.5,
+                pos="27,18",
+                width=0.75];
+            a:b -> c    [pos="e,27,36.104 27,71.697 27,63.983 27,54.712 27,46.112"];
+        }
+
+    ...in other words, our `a:b` node statements became just `a`, and our `a:b -> c`
+    edge statement gave us a warning that node "a" had no port "b".
+
+    Also, https://graphviz.org/doc/info/lang.html says:
+
+        Note also that the allowed compass point values are not keywords, so these
+        strings can be used elsewhere as ordinary identifiers and, conversely, the
+        parser will actually accept any identifier.
+
+    ...and experimentally, you can determine at commandline that "any identifier" here
+    means "any ID", where ID is a possibly-quoted value.
+    For example, the compass value "x y" is accepted: `echo 'digraph { a:b:"x y" -> c }' | dot`
+    (We are therefore justified in calling quote() on self.compass in NodeID.__str__.)
+    """
+
+    def __init__(self, id, port='', compass=''):
+        # NOTE: port and compass should indeed default to '', not None.
+        # The DOT semantics do not distinguish between their missing vs being provided
+        # as (quoted) empty strings.
+        # For instance, this is a syntax error:
+        #     echo 'digraph { a: -> c }' | dot
+        # ...so is this:
+        #     echo 'digraph { a::n -> c }' | dot
+        # ...whereas the following is accepted, and transformed into "a -> c":
+        #     echo 'digraph { a:"" -> c }' | dot
+        self.id = escape(id)
+        self.port = escape(port)
+        self.compass = escape(compass)
+
+    def __str__(self):
+        s = quote(self.id)
+        if self.port:
+            s = f'{s}:{quote(self.port)}'
+        if self.compass:
```

Comment:
Did you consider creating a `list` and using `':'.join()`?

Reply:
No, I find this equally readable. Obviously either way is equivalent, so I wouldn't mind if it were switched to the other way.